access/resource_table_acl.go

```diff
@@ -0,0 +1,339 @@
+}
+
+func (ta *TableACL) permissions() map[string][]TablePermissions {
+	return map[string][]TablePermissions{
+		"GRANT": ta.Grants,
+		"DENY":  ta.Denies,
+	}
+}
+
+func (ta *TableACL) actualDatabase() string {
+	if ta.Database == "" {
+		return "default"
+	}
+	return ta.Database
+}
+
+// TypeAndKey returns ACL object type and key
+func (ta *TableACL) TypeAndKey() (string, string) {
+	if ta.Table != "" {
+		return "TABLE", fmt.Sprintf("`%s`.`%s`", ta.actualDatabase(), ta.Table)
+	}
+	if ta.View != "" {
+		return "VIEW", fmt.Sprintf("`%s`.`%s`", ta.actualDatabase(), ta.View)
+	}
+	if ta.Database != "" {
+		return "DATABASE", ta.Database
+	}
+	if ta.Catalog {
+		return "CATALOG", ""
+	}
+	if ta.AnyFile {
+		return "ANY FILE", ""
+	}
+	if ta.AnonymousFunction {
+		return "ANONYMOUS FUNCTION", ""
+	}
+	return "", ""
+}
+
+// ID returns Terraform resource ID
+func (ta *TableACL) ID() string {
+	objectType, key := ta.TypeAndKey()
+	noBackticks := strings.ReplaceAll(key, "`", "")
+	return fmt.Sprintf("%s/%s", strings.ToLower(objectType), noBackticks)
+}
+
+func loadTableACL(id string) (TableACL, error) {
+	ta := TableACL{}
+	split := strings.SplitN(id, "/", 2)
+	if len(split) != 2 {
+		return ta, fmt.Errorf("ID must be three elements: %s", id)
+	}
+	switch strings.ToLower(split[0]) {
+	case "database":
+		ta.Database = split[1]
+	case "view":
+		dav := strings.SplitN(split[1], ".", 2)
+		if len(dav) != 2 {
+			return ta, fmt.Errorf("View must be two elements")
+		}
+		ta.Database = dav[0]
+		ta.View = dav[1]
+	case "table":
+		dav := strings.SplitN(split[1], ".", 2)
+		if len(dav) != 2 {
+			return ta, fmt.Errorf("Table must be two elements")
+		}
+		ta.Database = dav[0]
+		ta.Table = dav[1]
+	case "catalog":
+		ta.Catalog = true
+	case "any file":
+		ta.AnyFile = true
+	case "anonymous function":
+		ta.AnonymousFunction = true
+	default:
+		return ta, fmt.Errorf("Illegal ID type: %s", split[0])
+	}
+	return ta, nil
+}
+
+func (ta *TableACL) read(exec common.CommandExecutor) error {
+	objectType, key := ta.TypeAndKey()
+	result := exec.Execute(ta.ClusterID, "sql", fmt.Sprintf(
+		"SHOW GRANT ON %s %s", objectType, key))
+	if result.Failed() {
+		failure := result.Error()
+		if strings.Contains(failure, "does not exist") ||
+			strings.Contains(failure, "RESOURCE_DOES_NOT_EXIST") {
+			return common.NotFound(failure)
+		}
+		return fmt.Errorf(failure)
+	}
+	// clear any previous entries
+	ta.Grants = []TablePermissions{}
+	ta.Denies = []TablePermissions{}
+	var principal, actionType, objType, objectKey string
+	for result.Scan(&principal, &actionType, &objType, &objectKey) {
+		if !strings.EqualFold(objectType, objType) {
+			continue
+		}
+		if !strings.EqualFold(objectKey, key) {
+			continue
+		}
+		target := ta.Grants
+		var isDeny = strings.HasPrefix(actionType, "DENIED_")
+		if isDeny {
+			actionType = strings.Replace(actionType, "DENIED_", "", 1)
+			target = ta.Denies
+		}
+		var privileges *[]string
```

Comment:
@alexott can you tell me if i'm not doing anything stupid here? you've worked with pointers more than me, so you know for sure :P

Reply:
from looking to the code, it looks ok...